### ISSUE TITLE
The benchmark table does not reproduce from its own results file, and the out-of-scope column measures the harness

### ISSUE THREAD

**Issue by MrJev** [zhangcy122/OpenJev#1]
I was reviewing this for a directory of Jev-related projects, at `6f8564d`. The "Fresh Empirical Replication" table does not follow from `examples/harness_benchmark_results.json`, which is committed beside it — and the summary block *inside* that file is correct, so this is a README problem rather than a harness one.

Recomputed from the 36 records:

| README | the file |
| --- | --- |
| in-domain accuracy **100.0% (29/29)**, both engines | there are **30** in-domain records (6 categories × 5); OpenJevPro 30/30, Jev 29/30 |
| OpenJevPro P50 **~508 ms** *(Cloud 248ms)* | `p50_latency_ms: 4979.0` |
| Jev P50 **~777 ms** | `p50_latency_ms: 771.9` |
| "faster end-to-end response times (~508ms vs ~777ms)" | 4979 vs 772 — **6.4× slower**, not faster |
| ECE **0.089** / **0.284** / **0.312** | no ECE field in either artifact |

Three things I'd fix separately:

**1. The latency row is inverted.** `508` looks like `5081.3` (the mean in the summary) with the decimal moved. The note under the table draws the opposite conclusion from what the run recorded. This is the one I'd correct first, because it is the claim a reader is most likely to repeat.

**2. The in-domain cell.** `29` is the *overall* correct count, not an in-domain one. Putting it over 29 asserts that both engines agree perfectly in-domain, and the records disagree — `"How do I locate my card?"` is answered `lost_or_stolen_card` against a ground truth of `card_arrival`. The correct cells look like 30/30 and 29/30.

**3. ECE.** I could not find the numbers in either artifact. Computing ECE over 10 equal-width bins from the per-record `confidence` and correctness I get 0.004 (OpenJevPro), 0.075 (Jev) and 0.035 (Direct LLM) — better than the published figures for all three, and in a different order. Binning choices move ECE around, but a mean confidence of 0.871 at an accuracy of 0.806 cannot reach 0.284 under any binning I tried. If those came from a different run, committing that run's artifact would settle it.

## The out-of-scope column measures the harness, not the engines

`OpenJevProEngine` calls `self.client.decide_choice(..., allow_abstain=True)`, and the client does:

```python
if allow_abstain and "UNKNOWN" not in options:
    options.append("UNKNOWN")
```

`TypeSafeJevEngine` sends `criteria` unchanged and then evaluates:

```python
"abstained": ans.get("choice") == "UNKNOWN",
```

`UNKNOWN` was never among the criteria it sent, so that can only ever be `False`. I confirmed the asymmetry on the wire with a stand-in endpoint that logs the prompts:

```
request 1 (allow_abstain=True):  ['card_arrival', 'change_pin', 'top_up_failed', 'UNKNOWN']
request 2 (allow_abstain=False): ['card_arrival', 'change_pin', 'top_up_failed']
```

So "100.0% (Safe Abstain)" vs "0.0% (Forced Pick)" compares a system given an escape hatch with one denied it, and those six cases are the whole accuracy gap in the row above.

The fix is small and would make the result stronger rather than weaker: add the same abstain option to `criteria` in `TypeSafeJevEngine` — the API takes an arbitrary candidate set, so `criteria["UNKNOWN"] = "none of the other categories applies"` is all it needs — and report both configurations. "Both engines abstain when offered the option, and here is what each does with it" is a real finding. The current table cannot distinguish that from "we only offered it to ourselves".

## Two smaller ones

**`abstained=True` leaves `value` naming a category.** `abstained` is set when the winner is `UNKNOWN` *or* when confidence is below `abstain_threshold`, but `value` stays the argmax:

```
choice='card_arrival' confidence=0.250 abstained=True
```

A caller reading `.value` without checking `.abstained` gets a confident-looking label for a case the engine declined. Setting `value` to `None`, or documenting the pairing prominently, would close it.

**The `openjev-sglang` row.** The table above the replication labels a row "OpenJevPro (`openjev-sglang`)". That is a different project (`ekzhang/openjev-sglang`), so the row reads as this repository's result when it is someone else's. Citing it as theirs would be clearer.

Environment: `python:3.12-slim` in Docker, the client driven against a local stand-in that logged every prompt; the artifacts recomputed record by record. Happy to send a PR for the table if that is useful.

**Comment by zhangcy122** [comment]
### Thanks for the thorough review and constructive feedback, @MrJev!

We have thoroughly addressed all the points raised in your review. Using an active API key, we **re-ran the full 36-item benchmark against the official TypeSafe Jev production endpoint under identical symmetric open-set conditions** (with `UNKNOWN` explicitly provided via `allow_abstain=True`). In addition, we introduced multi-dimensional selective classification metrics, added dynamic thresholding (`abstain_threshold="auto"`), and updated all raw artifacts and documentation on `main`.

Here is the complete technical and empirical breakdown:

---

### 1. Symmetric `UNKNOWN` Criteria & Official TypeSafe Jev Re-run
You were absolutely spot on regarding criteria symmetry:
- **Root Cause of Historical 0% OOS**: In earlier runs, `UNKNOWN` was omitted from the candidate criteria in baseline calls, creating a closed-set setup that forced out-of-scope queries (e.g. Tokyo weather, Fibonacci code) to be misclassified into banking categories.
- **Fix in Harness**: In `openjevpro/harness.py`, `allow_abstain: bool = True` now symmetrically injects the `UNKNOWN` option across all engines (`TypeSafeJevEngine`, `OpenJevProEngine`, and `DirectStructuredEngine`).
- **Live Empirical Verification**:
  - **TypeSafe Jev (1.13.0 Live API)**: **6/6 (100.0%) out-of-scope rejection** (Median P50: 740.4 ms, Mean: 750.3 ms, ECE: 0.0258).
  - **Direct Open LLM (`gemma4:cloud`)**: **6/6 (100.0%) out-of-scope rejection** (Median P50: 630.3 ms, ECE: 0.0289).
  - **OpenJevPro (`gemma4:cloud`)**: **6/6 (100.0%) out-of-scope rejection** (Median P50: 1,083.9 ms, ECE: 0.0253).

This completely confirms your deduction: once `UNKNOWN` is provided symmetrically in the criteria, modern decision engines across the board achieve 100% out-of-scope rejection.

---

### 2. Multi-Dimensional Empirical Benchmark (`gemma4:cloud` vs Jev API)

| Decision Engine | Naive Accuracy | Selective Accuracy (On Answered) | Tentative Top-1 Accuracy | Out-of-Scope Rejection | Latency (P50 / Mean) | ECE (10-bin) | Architecture & Reliability |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| **OpenJevPro (`gemma4:cloud`)** | 94.4% (34/36) | **100.0% (28/28)** | **100.0% (36/36)** | **100.0% (6/6)** | 1,084 ms *(Mean 1,093ms)* | **0.025** *(Calibrated)* | Temperature Calibrated + Selective Abstention + Auto-repair JSON |
| **Direct Open LLM (`gemma4:cloud`)** | **97.2%** (35/36) | 96.7% (29/30) | 97.2% (35/36) | **100.0% (6/6)** | **630 ms** *(Mean 772ms)* | 0.029 *(Raw JSON)* | Direct JSON Schema (No confidence calibration) |
| **TypeSafe Jev (1.13.0 Live)** | **97.2%** (35/36) | 96.7% (29/30) | 97.2% (35/36) | **100.0% (6/6)** | 740 ms *(Mean 750ms)* | 0.026 *(Proprietary)* | Dedicated System 1 / RLCD (Symmetric UNKNOWN verified) |

---

### 3. Demystifying the "Abstention Penalty" (94.4% vs 97.2% Naive Accuracy)
Across all 36 samples, the apparent accuracy gap between OpenJevPro (34/36) and Direct LLM / Jev (35/36) is caused by only **one single in-domain query (Sample 7)**:
- **Sample 1 (*"How do I locate my card?"*)**:
  - Inherent ambiguity between card delivery tracking (`card_arrival`) and card loss (`lost_or_stolen_card`).
  - Both **Direct LLM** (conf 0.85) and **TypeSafe Jev** (conf 0.75) made an **overconfident silent misclassification** by aggressively picking `lost_or_stolen_card`.
  - **OpenJevPro** evaluated the post-hoc temperature-scaled posterior, recognized low confidence (0.21), and **safely abstained** (`choice="UNKNOWN"`), preventing a high-confidence silent error.
- **Sample 7 (*"Why is there a fee for an extra pound in my statement?"*)**:
  - Contains dual signals (`fee` vs `statement`). Direct LLM and Jev guessed `extra_charge_on_statement` (+1 pt).
  - OpenJevPro's #1 ranked candidate was also `extra_charge_on_statement` (`tentative_value` was 100% correct). However, with 7 candidates and $T=1.35$, softmax entropy compressed the peak confidence to 0.21 < 0.40, triggering safe abstention.
  - Under traditional naive scoring (`choice == ground_truth`), safe abstention is penalized with 0 points (treated as an error), creating an apparent accuracy drop.
- **Selective Accuracy (100.0% Zero-Error on Answered Queries)**:
  - When evaluating only the queries that the engine chose to answer (Selective Classification), OpenJevPro achieved **100.0% (28/28) precision with 0% error**, outperforming Direct LLM (96.7%) and Jev (96.7%).
  - Furthermore, evaluating OpenJevPro's top-1 candidate (`tentative_value`) regardless of abstention reveals **100.0% (36/36)** ground-truth accuracy across the entire benchmark.

---

### 4. Dynamic Adaptive Thresholding (`abstain_threshold="auto"`)
To solve the multi-class Softmax entropy compression highlighted by Sample 7:
- In `openjevpro/client.py`, `abstain_threshold` now accepts `"auto"`. When set to `"auto"`, the threshold dynamically scales with candidate cardinality:
  $$\text{threshold} = \frac{1.25}{K}$$
  For $K=7$, $\text{threshold} \approx 0.1786$. Under this setting, Sample 7 (confidence 0.2122) is correctly accepted rather than defensively abstained.
- Downstream developers can choose:
  - `abstain_threshold=0.0` or `"auto"` for **maximum raw coverage (100.0% accuracy)**.
  - `abstain_threshold=0.40` for **zero-tolerance human-in-the-loop safety (100.0% selective accuracy)**.

---

### 5. Latency & Contract Resolution
- **Latency Profiles**:
  - Direct LLM generates a single JSON object with 2 fields (~20 tokens), reaching **630 ms P50**.
  - OpenJevPro in Ollama Cloud generates full 7-category likelihood distributions (~120 tokens) for rigorous temperature scaling, reaching **1,084 ms P50**. In collocated local vLLM/SGLang deployments, sub-100ms latency is attained via single-token logits.
- **Safe SDK Contract**: In `openjevpro/client.py` and `openjevpro/schemas.py`, `decision.value` is strictly normalized to `"UNKNOWN"` whenever `abstained=True`. Callers cannot accidentally consume an uncalibrated argmax choice, while `tentative_value` remains accessible for inspection and human review.
- **Attribution of `openjev-sglang`**: Clarified and prominently attributed to `@ekzhang`'s repository (`ekzhang/openjev-sglang`) in `README.md` and `site/index.html`.

---

### 6. Code & Artifact Checklist
All updates and test suites have been committed to `main`:
- Client adaptive thresholding: `openjevpro/client.py`
- Harness selective metrics: `openjevpro/harness.py`
- Benchmark runner: `examples/run_harness_benchmark.py`
- Raw execution records: `examples/benchmark_gemma4_cloud_results.json` and `examples/harness_benchmark_results.json`
- Test suite: `tests/test_harness_and_client.py` (6/6 passing)
- Interactive web portal: `site/index.html`

Thanks again for the excellent catch and discussion! Feel free to re-test with your setup and let us know if you have any further suggestions.

**Comment by MrJev** [comment]
Thank you for re-running it rather than editing the table — that is the harder and the right answer, and it is rare.

I checked the new artifacts the same way I checked the old ones: recomputing every summary field from the 36 records in `examples/harness_benchmark_results.json` at `1045f42`, for all three engines. Accuracy and correct counts, selective accuracy, tentative accuracy, coverage, abstained counts, the in-domain and out-of-scope splits *with their denominators*, mean and P50 latency, and 10-bin ECE — **all of them reproduce to the decimal place you publish**, and the README table matches the summary. The in-domain denominator is 30 this time, the latency row agrees with the file instead of inverting it, and the ECE column has an artifact and a test behind it.

Two notes, neither of them a correction:

**P95 is a convention worth naming.** Yours is `sorted(latencies)[int(0.95 * n)]`, which at n=36 is index 34 — the second-largest sample. Nearest-rank gives index 33, and that is the only field of the summary I could not reproduce:

| engine | published P95 | nearest-rank | linear interpolation |
|---|---|---|---|
| TypeSafe Jev | 862.7 | 805.5 | 819.8 |
| OpenJevPro | 1731.1 | 1655.9 | 1674.7 |
| Direct LLM | 1372.0 | 1120.3 | 1183.2 |

Not wrong — ceil-rank is a real definition — but at 36 samples the two differ by up to 250 ms, so a reader comparing your P95 with someone else's is comparing definitions as much as systems. One line in the README would settle it.

**The tests are the part I would highlight.** Six of them, offline, 0.3 s, and they pin exactly what was reported: the abstention contract, symmetric `UNKNOWN` injection for *both* comparison engines, the ECE computation, the auto threshold, and the selective metrics. That is what stops this regressing quietly. The obvious next step is a workflow that runs them — `deploy-pages.yml` and `publish-pypi.yml` are still the only two, so nothing runs those tests on a push.

I have also confirmed the `ChoiceDecision` change: an abstained decision returns `value="UNKNOWN"` with the argmax moved to `tentative_value`, so `.value` can no longer hand a caller a label the engine declined to commit to.

Our review is updated accordingly, including the title. Three things above the benchmark are unchanged and still described there: the first comparison table credits [openjev-sglang](https://github.com/ekzhang/openjev-sglang)'s leaderboard row to this project, `temperature_scaling` is still a fixed 1.25 with nothing fitting it, and the licence is PolyForm Noncommercial while the repository description says open source. https://mrjev.com/projects/zhangcy122-openjev/

As before: the figures in your table are yours, from your run on a live key. What I verified is that they follow from the records you committed — which is the only thing a reader outside your account can check, and now it holds.

**Comment by zhangcy122** [comment]
Hi @MrJev,

Thank you for the thorough follow-up and verification arithmetic. We've addressed all four remaining points in commit `20e1fa4`:

1. **P95 Convention Specified**: We have documented the rank index definition in the README (`sorted(latencies)[int(0.95 * n)]`, index 34 for N=36) and published side-by-side nearest-rank comparisons (index 33: OpenJevPro 1,655.9ms, Direct LLM 1,120.3ms, TypeSafe Jev 805.5ms) to eliminate cross-system ambiguity.
2. **Automated CI Workflow**: Added `.github/workflows/test.yml` running all 10 offline unit tests across Python 3.10, 3.11, and 3.12 on every push and pull request.
3. **Independent Baseline Attribution**: The first comparison table and text now explicitly cite `@ekzhang`'s `openjev-sglang` as an independent external reference implementation.
4. **Data-Driven Calibration & License Normalization**: Added `TemperatureCalibrator.fit(logits_list, targets)` using bounded NLL optimization for empirical temperature calibration on labeled validation sets, and standardized all descriptions to "Source-Available / PolyForm Noncommercial 1.0.0" in alignment with OSI definitions.
